### PR TITLE
VXQUERY-209 Parallel Index creation and access implementation

### PR DESCRIPTION
**Pull request by menakaj**

### CODE REVIEW

File: vxquery-core/src/main/java/org/apache/vxquery/common/VXQueryCommons.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    public static final Set<FunctionIdentifier> indexingFunctions = new HashSet<>();
+
+    public static IndexCentralizerUtil INDEX_CENTRALIZER_UTIL = null;
+    static {
+        INDEX_CENTRALIZER_UTIL = new IndexCentralizerUtil();
+        INDEX_CENTRALIZER_UTIL.readIndexDirectory();
+
+    }
+
+    static {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/common/VXQueryCommons.java:36]: Split this static assignment into two parts. Add each part below the variable its assigning values to. Similar to INDEX_CENTRALIZER_UTIL.

File: vxquery-core/src/main/java/org/apache/vxquery/common/VXQueryCommons.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+public class VXQueryCommons {
+
+    public static final Set<FunctionIdentifier> collectionFunctions = new HashSet<>();
+
+    public static final Set<FunctionIdentifier> indexingFunctions = new HashSet<>();
+
+    public static IndexCentralizerUtil INDEX_CENTRALIZER_UTIL = null;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/common/VXQueryCommons.java]: why declare it null instead of "new IndexCentralizerUtil()"?

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java [hunk 3]
```diff
@@ -16,76 +16,16 @@
...
-import org.apache.vxquery.compiler.rewriter.rules.RemoveRedundantCastExpressionsRule;
-import org.apache.vxquery.compiler.rewriter.rules.RemoveRedundantDataExpressionsRule;
-import org.apache.vxquery.compiler.rewriter.rules.RemoveRedundantPromoteExpressionsRule;
-import org.apache.vxquery.compiler.rewriter.rules.RemoveRedundantTreatExpressionsRule;
-import org.apache.vxquery.compiler.rewriter.rules.RemoveUnusedSortDistinctNodesRule;
-import org.apache.vxquery.compiler.rewriter.rules.RemoveUnusedUnnestIterateRule;
-import org.apache.vxquery.compiler.rewriter.rules.ReplaceSourceMapInDocExpression;
-import org.apache.vxquery.compiler.rewriter.rules.SetCollectionDataSourceRule;
-import org.apache.vxquery.compiler.rewriter.rules.SetVariableIdContextRule;
+import org.apache.hyracks.algebricks.core.rewriter.base.HeuristicOptimizer;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java:19]: use the hyracks profile to format the imports.
- alpha sorted
- no stars

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceIndexingRule.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+            // Build the new operator and update the query plan.
+            int collectionId = vxqueryContext.newCollectionId();
+            VXQueryIndexingDataSource ids = VXQueryIndexingDataSource.create(collectionId, collection, null,
+                    SequenceType.create(AnyItemType.INSTANCE, Quantifier.QUANT_STAR),
+                    functionCall.getFunctionIdentifier().getName());
+            if (ids != null) {
+                ids.setTotalDataSources(vxqueryContext.getTotalDataSources());
+
+                // Known to be true because of collection name.
```
Comment (ecarm002) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceIndexingRule.java]: extract the following lines that change the query plan to the abstract class. (share this code between the two rules.)

File: vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryIndexingDataSource.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    private String[] indexPartitions;
+    private int totalDataSources;
+    private String tag;
+    private String function;
+
+    private final Object[] types;
+
+    private IDataSourcePropertiesProvider propProvider;
+
+    private VXQueryIndexingDataSource(int id, String collection, String index, Object[] types, String functionCall) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryIndexingDataSource.java:38]: Create an AbstractVXQueryDataSouce to share information between indexing and collection. Start to clean up the code.

File: vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryIndexingOperatorDescriptor.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+import java.nio.ByteBuffer;
+import java.security.NoSuchAlgorithmException;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+public class VXQueryIndexingOperatorDescriptor extends AbstractSingleActivityOperatorDescriptor {
+    protected static final Logger LOGGER = Logger.getLogger(VXQueryCollectionOperatorDescriptor.class.getName());
+    private static final long serialVersionUID = 1L;
+    private final String START_TAG = "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n";
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryIndexingOperatorDescriptor.java]: Clean up variable list, some are not used.

File: vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryIndexingOperatorDescriptor.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+                        if (collectionDirectory.exists() && collectionDirectory.isDirectory()) {
+                            IndexConstructorUtil indexConstructorUtil = new IndexConstructorUtil();
+                            try {
+                                indexConstructorUtil
+                                        .evaluate(collectionModifiedName, indexModifiedName, result, stringp, bbis, di,
+                                                sb, abvs, nodeIdProvider, abvsFileNode, nodep, false, nodeId);
+                                XDMConstants.setTrue(result);
+                                FrameUtils.appendFieldToWriter(writer, appender, result.getByteArray(), result.getStartOffset(),
+                                        result.getLength());
+                            } catch (SystemException | JAXBException e) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryIndexingOperatorDescriptor.java]: Please clean up exception handling. Lets try to throw fewer errors. same for all index methods.

File: vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryIndexingOperatorDescriptor.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+            public void close() throws HyracksDataException {
+                // Check if needed?
+                if (appender.getTupleCount() > 0) {
+                    appender.flush(writer, true);
+                }
+                writer.close();
+                try {
+                    VXQueryCommons.INDEX_CENTRALIZER_UTIL.writeIndexDirectory();
+                } catch (JAXBException | FileNotFoundException e) {
+                    e.printStackTrace();
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryIndexingOperatorDescriptor.java]: throw error. Also consider better exceptions from writeIndexDirectory

File: vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryMetadataProvider.java [hunk 9]
```diff
@@ -88,22 +94,51 @@ public StaticContext getStaticContext() {
...
+
+
+        //Remove following redundant code
+        IOperatorDescriptor scanner = null;
+        AlgebricksPartitionConstraint constraint = null;
+        if (VXQueryCommons.indexingFunctions.contains(
+                AbstractCollectionRule.functionCall.getFunctionIdentifier())) {
+            //Indexing
+            VXQueryIndexingDataSource ds = (VXQueryIndexingDataSource) dataSource;
+            if (sourceFileMap != null) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryMetadataProvider.java:106]: Can indexing and collection share any of this code in the if block?

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexConstructorScalarEvaluatorFactory.java [hunk 10]
```diff
@@ -65,11 +68,27 @@ protected IScalarEvaluator createEvaluator(IHyracksTaskContext ctx, IScalarEvalu
...
+                            stringp.getLength() + stringp.getStartOffset())), 0);
+                    indexFolder = di.readUTF();
+
+                    IndexConstructorUtil indexConstructorUtil = new IndexConstructorUtil();
+                    indexConstructorUtil.evaluate(collectionFolder, indexFolder, result, stringp, bbis, di, sb, abvs,
+                            nodeIdProvider, abvsFileNode,
                             nodep, false, nodeId);
                     XDMConstants.setTrue(result);
-                } catch (JAXBException e) {
+                } catch (JAXBException | IOException e) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexConstructorScalarEvaluatorFactory.java]: don't throw the JAXBException error.

File: vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml [hunk 11]
```diff
@@ -166,6 +165,12 @@
         </property>
     </function>
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml:165]: Remember to remove the old runtime references.

**Comment by menakaj** [comment]
I did the changes according to the code review and also completed the user documentation.
- The issue with show-index output is fixed.
- Cleaned up the code and exceptions.

Issues.
The results of collection-from-index query are not printed to the console. I'm working on fixing it.

File: src/site/apt/user_indexing.apt [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+*1. Scenario I - When collection is a single directory.
+In this scenario, all the XML files are stored in a single directory. (There can be sub directories)
+
+*** Creating an index for collection
+If I need to create index for xml collection stored <path_to_collection>/collection1,
+
+Query structure:
+--------
+build-index-on-collection("<path_to_collection>")
```
Comment (sjaco002) [inline src/site/apt/user_indexing.apt]: Should be <path_to_collection>/collection1

File: src/site/apt/user_indexing.apt [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+--------
+
+You can see the index has created in a new sub-directory in the index_directory specified in local.xml
+
+Example:
+--------
+build-index-on-collection("<path1>/collection1")
```
Comment (sjaco002) [inline src/site/apt/user_indexing.apt]: Leave it the same < path_to_collection >/collection1

File: src/site/apt/user_indexing.apt [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+[[b]] Use the index in executing a query.
+[[c]] Update the index.
+[[d]] Delete the index.
+[[e]] View existing indexes.
+
+*1. Scenario I - When collection is a single directory.
+In this scenario, all the XML files are stored in a single directory. (There can be sub directories)
+
+*** Creating an index for collection
+If I need to create index for xml collection stored <path_to_collection>/collection1,
```
Comment (sjaco002) [inline src/site/apt/user_indexing.apt:43]: Change this to < path1 >/collection1 to match the example queries rather than the structure queries

File: src/site/apt/user_indexing.apt [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+This function also takes the collection path of which the index is needed to be deleted.
+
+Query structure:
+--------
+delete-index("<path_to_collection>")
+--------
+
+Example:
+-------
+delete-index("<path_to_collection>/collection1")
```
Comment (sjaco002) [inline src/site/apt/user_indexing.apt:112]: < path1 >/collection1

File: src/site/apt/user_indexing.apt [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+-------
+
+*2. Scenario II - When the collection is distributed.
+In this scenario, the collection is distributed among several partitions.
```
Comment (sjaco002) [inline src/site/apt/user_indexing.apt]: In this scenario, the collection is distributed among several directories. We can distribute the queries among partitions

File: src/site/apt/user_indexing.apt [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+Query structure:
+--------
+build-index-on-collection("<partition_1_path>|<partition_2_path>|<partition_3_path>|...|<partition_n_path>")
+--------
+
+In here the parameter contains the list of collection partitions separated by '|' character.
+
+Example:
+Consider the collection has now distributed among four partitions, <path_1>/collection1, <path_2>/collection2,
```
Comment (sjaco002) [inline src/site/apt/user_indexing.apt]: "among four directories"

File: src/site/apt/user_indexing.apt [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+<path_3>/collection3 and <path_4>/collection4.
+
+To create indexes for all of the above collections,
+-------
+build-index-on-collection("<path_1>/collection1|<path_2>/collection2|<path_3>/collection3|<path_4>/collection4")
+-------
+
+In this case, all indexes will be created in separate sub-directories corresponding to each partition.
```
Comment (sjaco002) [inline src/site/apt/user_indexing.apt]: Also note that this query requires each node to have four partitions available

File: src/site/apt/user_indexing.apt [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+indexes of those partitions.
+
+In this case, give a '|' separated list of collection partitions.
+Query structure:
+--------
+update-index("<partition_1_path>|<partition_2_path>|<partition_3_path>|...|<partition_n_path>")
+--------
+
+Example:
+Suppose that we need to update the indexes in partition1 and partition4
```
Comment (sjaco002) [inline src/site/apt/user_indexing.apt:168]: "update the indexes for collection1 and collection4"

File: src/site/apt/user_indexing.apt [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+
+*** Deleting the indexes.
+If we want to delete indexes of collections in several partitions, we can use this function.
+Query structure:
+--------
+delete-index("<partition_1_path>|<partition_2_path>|<partition_3_path>|...|<partition_n_path>")
+--------
+
+Example:
+Suppose that we need to update the indexes in partition2 and partition3
```
Comment (sjaco002) [inline src/site/apt/user_indexing.apt]: "update the indexes for collection2 and collection3"

File: src/site/apt/user_indexing.apt [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+--------
+delete-index("<path_1>/collection2|<path_4>/collection3")
```
Comment (sjaco002) [inline src/site/apt/user_indexing.apt]: The paths are wrong here

File: vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryMetadataProvider.java (hunk 9 above)
Comment (sjaco002) [inline vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryMetadataProvider.java:106]: +1

File: vxquery-server/src/main/resources/conf/local.xml [hunk 22]
```diff
@@ -16,6 +16,7 @@
 -->
 <cluster xmlns="cluster">
     <name>local</name>
+    <indexDirectory>/home/menaka/index</indexDirectory>
```
Comment (sjaco002) [inline vxquery-server/src/main/resources/conf/local.xml]: I think by default this should be the directory created in AbstractXQueryTests. Then you can create a bigger set of queries in src/test/resources/Queries/Indexing. It seems to me that these tests will fail currently on this branch, as they haven't been changed yet.

Comment (menakaj) [inline vxquery-server/src/main/resources/conf/local.xml]: Yes. This was also  a problem that I had also because of the XTests. I think this solution will work fine. So, I'll keep it as target/tmp/indexFolder

File: vxquery-core/src/main/java/org/apache/vxquery/common/VXQueryCommons.java [hunk 23]
```diff
@@ -23,18 +23,13 @@
 
 import javax.xml.bind.JAXBException;
 import javax.xml.parsers.ParserConfigurationException;
+import java.io.File;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/common/VXQueryCommons.java]: Why add java.io.File? Make sure its a clean list of imports.

File: vxquery-xtest/src/test/resources/Queries/XQuery/Indexing/useIndex7.xq [hunk 24]
```diff
@@ -18,8 +18,8 @@
 (: Search Lucene Index :)
 (: Find all the weather readings for Washington state for a specific day    :)
 (: 2002-2-2.                                                                  :)
-for $s in collection-from-index("target/tmp/indexFolder", "/stationCollection/station")/station
-for $r in collection-from-index("target/tmp/indexFolder", "/dataCollection/data")/data
+for $s in collection-from-index("src/test/resources/TestSources/ghcnd", "/stationCollection/station")/station
```
Comment (prestoncarman) [inline vxquery-xtest/src/test/resources/Queries/XQuery/Indexing/useIndex7.xq:21]: Can you use the sources key similar to the GHCND queries instead of the file path? You can see the map of sources in the top of the VXQueryCatalog.xml file.

Comment (prestoncarman) [inline vxquery-xtest/src/test/resources/Queries/XQuery/Indexing/useIndex7.xq:21]: This goes for all the index tests.

File: vxquery-xtest/src/main/java/org/apache/vxquery/xtest/TestClusterUtil.java [hunk 25]
```diff
@@ -34,7 +34,8 @@
     private static int profileDumpPeriod = 10000;
     private static String ccHost = "localhost";
     private static String nodeId = "nc1";
-
+//    private static String ioDevices = "target/tmp/indexFolder";
```
Comment (prestoncarman) [inline vxquery-xtest/src/main/java/org/apache/vxquery/xtest/TestClusterUtil.java]: Were you able to get this path to work?

File: src/site/apt/user_indexing.apt [hunk 26]
```diff
@@ -44,14 +44,14 @@ If I need to create index for xml collection stored <path_to_collection>/collect
 
 Query structure:
 --------
-build-index-on-collection("<path_to_collection>")
+build-index-on-collection("collection")
```
Comment (prestoncarman) [inline src/site/apt/user_indexing.apt:47]: How about using "`<path_to_collection>`" or "`<path_to_collection_1>`" and "`<path_to_collection_2>`"?

The example paths in this document are not consistent.

Comment (menakaj) [inline src/site/apt/user_indexing.apt:47]: For examples, may I use the actual test resource paths?

File: vxquery-xtest/src/test/resources/Queries/XQuery/Indexing/useIndex7.xq (hunk 24 above)
Comment (menakaj) [inline vxquery-xtest/src/test/resources/Queries/XQuery/Indexing/useIndex7.xq:21]: I tried it but then, for results it returns the absolute path.

File: vxquery-benchmark/src/main/resources/noaa-ghcn-daily/queries_json/no_result.xq [hunk 27]
```diff
@@ -1,25 +0,0 @@
...
-   KIND, either express or implied.  See the License for the
-   specific language governing permissions and limitations
-   under the License. :)
-
-(: 
-VXQuery used to only parse all files with out producing results.
-:)
-let $sensor_collection := "/tmp/1.0_partition_ghcnd_all_xml/sensors"
-for $r in collection($sensor_collection)
-for $data in $r("dataCollection")("data")
```
Comment (sjaco002) [inline vxquery-benchmark/src/main/resources/noaa-ghcn-daily/queries_json/no_result.xq:23]: why is this deleted?

File: vxquery-cli/src/main/java/org/apache/vxquery/cli/VXQuery.java [hunk 28]
```diff
@@ -305,7 +305,7 @@ public void startLocalHyracks() throws Exception {
             ncConfig.dataIPAddress = localAddress;
             ncConfig.resultIPAddress = localAddress;
             ncConfig.nodeId = "nc" + (i + 1);
-            ncConfig.ioDevices = Files.createTempDirectory(ncConfig.nodeId).toString();
+            ncConfig.ioDevices = "target/tmp/indexFolder";
```
Comment (sjaco002) [inline vxquery-cli/src/main/java/org/apache/vxquery/cli/VXQuery.java:308]: For now let's leave this as it was. Not sure if the change would work in all scenarios

File: vxquery-server/src/main/resources/conf/local.xml [hunk 29]
```diff
@@ -16,6 +16,7 @@
 -->
 <cluster xmlns="cluster">
     <name>local</name>
+    <index_directory></index_directory>
```
Comment (sjaco002) [inline vxquery-server/src/main/resources/conf/local.xml]: go ahead and put a folder here /tmp so that there is a valid value example

File: vxquery-server/src/main/resources/conf/cluster_example.xml [hunk 30]
```diff
@@ -17,6 +17,7 @@
 <cluster xmlns="cluster">
     <name>local</name>
     <username>joe</username>
+    <index_directory></index_directory>
```
Comment (sjaco002) [inline vxquery-server/src/main/resources/conf/cluster_example.xml]: go ahead and put a folder here /tmp so that there is a valid value example

File: src/site/apt/user_indexing.apt (hunk 14 above)
Comment (sjaco002) [inline src/site/apt/user_indexing.apt:43]: Actually, let's replace the use of two steps `<path_to_collection>/collection1` with a single variable, e.g. `<path_to_collection1>`

File: vxquery-benchmark/src/main/resources/noaa-ghcn-daily/queries_json/no_result.xq (hunk 27 above)
Comment (menakaj) [inline vxquery-benchmark/src/main/resources/noaa-ghcn-daily/queries_json/no_result.xq:23]: Seems like I have pulled from the wrong branch. I'll restore this.
